packages/cli/src/commands/auth/login/loginFlow.ts

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2023 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// eslint-disable-next-line import/no-named-as-default
+import consola from "consola";
+import { createServer } from "http";
+import open from "open";
+import { join } from "path/posix";
+import { parse } from "url";
+import type { LoginArgs } from "./LoginArgs.js";
+
+export default async function invokeLoginFlow(args: LoginArgs) {
+  consola.start(`Authenticating using application id: ${args.applicationId}`);
+  const redirectUrl = "http://localhost:8080/auth/callback";
+  const port = parse(redirectUrl).port;
+  let resolve: (value: string) => void;
+  const authCode: Promise<string> = new Promise((_res) => {
+    resolve = _res;
+  });
+
+  const server = createServer((req, res) => {
+    const query = parse(req.url!, true).query;
+    res.end("Authenticated");
+    resolve(query["code"] as string);
+  });
+
+  server.listen(port);
```

Comment:
Probably want to give a clean error message if the port is already in use by the app's webpack or whatever